### DISCUSSION TITLE
chown: changing ownership of '/var/lib/mysql/': Permission denied on CentOS 9

### DISCUSSION THREAD

**Discussion by bhavers** [photoprism/photoprism#3955] [Q&A]
Hi,
I have PhotoPrism installed in CentOS 9 on Proxmox using podman.
When i do `podman-compose up -d` (using a non-root user that is part of the wheel group) mariadb and photoprism containers are started. But the log `podman-compose logs --tail=100 -f` repeately reports:

```
c86ae0dc6fea 2023-12-15 16:52:02+00:00 [Note] [Entrypoint]: Entrypoint script for MariaDB Server 1:10.11.6+maria~ubu2204 started.
c86ae0dc6fea find: ‘/var/lib/mysql/’: Permission denied
c86ae0dc6fea chown: changing ownership of '/var/lib/mysql/': Permission denied
c86ae0dc6fea 2023-12-15 16:52:02+00:00 [Note] [Entrypoint]: Entrypoint script for MariaDB Server 1:10.11.6+maria~ubu2204 started.
c86ae0dc6fea find: ‘/var/lib/mysql/’: Permission denied
```
When i enable `user: 1000:1000` in docker-compose.yml the log returns:
```
78baca0e6ae0 2023-12-15 16:44:52+00:00 [Note] [Entrypoint]: Entrypoint script for MariaDB Server 1:10.11.6+maria~ubu2204 started.
78baca0e6ae0 2023-12-15 16:44:52+00:00 [Note] [Entrypoint]: Initializing database files
78baca0e6ae0 2023-12-15 16:44:52 0 [Warning] Can't create test file '/var/lib/mysql/78baca0e6ae0.lower-test' (Errcode: 13 "Permission denied")
78baca0e6ae0 2023-12-15 16:44:52 0 [ERROR] mariadbd: Can't create/write to file './ddl_recovery.log' (Errcode: 13 "Permission denied")
78baca0e6ae0 2023-12-15 16:44:52 0 [ERROR] DDL_LOG: Failed to create ddl log file: ./ddl_recovery.log
78baca0e6ae0 2023-12-15 16:44:52 0 [ERROR] Aborting
```
When starting with `sudo` i get the same result. 
When i change the mariadb volume from /var/lib/mysql to something else (below), it starts normal (but i guess the data is than not persisted outside the container).
```
volumes:
      - "/home/user/database:/home/user/data"
```
 
Any suggestions that can help me with this last step to get it working? 
Appreciated, thx.

**Comment by lastzero** [comment]
Did you check the permissions and owner of the database storage directory you are trying to mount? Some red hat users also report that they need to mount volumes with the :z modifier at the end.

**Reply by bhavers** [reply]
I've `chmod 777` that `/home/user/database` directory the user is the owner. 
The problem seems to be with permission on /var/lib/mysql inside the container. I am not sure what i can about that.
I've tried adding the :z modifier, but that mostly produced all kinds of errors. Although i also don't understand how to apply that (tried various combination). I did see this: `https://docs.docker.com/storage/bind-mounts/#configure-the-selinux-label`. If you have other tips on this that would be great. thx.

**Reply by mthome1** [reply]
I had the same problem. After I started the container I did a chown -R user:gid dbDirectory as root. I had to do that after the first start only (with docker yml  restart: unless-stopped). It's probably not the most elegant solution but it fixed my problem for now :-D

**Reply by lastzero** [reply]
Note that we have also updated our Podman configuration examples based on your reports:
- https://github.com/photoprism/photoprism/commit/ef776a1a465546b13be2c56d8b9fc97cef534e9a

However, this does not help if you have manually changed the user of the mariadb service, while the related storage folder was created as root. In this case, you will need to change the folder permissions and restart the service manually.

**Reply by bhavers** [reply]
Thanks, unfortunately i still  don't get it to work. I tried chown -R (also made it readable, writable and executable for all) and updated the docker-compose.yml file with security_opt enabled.

**Reply by lastzero** [reply]
Did you check mariadb / red hat specific forums or Google for advice yet? I would think there's a solution for that...

**Reply by bhavers** [reply]
Yes did so last week, but have to do it again. Will let you know if i get it running.

**Reply by bhavers** [reply]
It definitely is something related to SELinux. When i disable SELinux `su -c "setenforce 0"` the permission problem is solved (and i do get data in that database directory). Though some other problem (locking) happens. Obviously disabling SELinux is not what i want. So that z modifier is probably the solution though i don't understand how it works. Have to investigate that later. Will let you know. Here's a link to relevant posts: 
https://github.com/docker-library/mysql/issues/198
https://stackoverflow.com/questions/24288616/permission-denied-on-accessing-host-directory-in-docker
